### DISCUSSION TITLE
Filter out dependencies versions?

### DISCUSSION THREAD

**Discussion by jmfayard** [Splitties/refreshVersions#349] [Ideas]
Hello friends, 
Did you ever want to filter out some dependencies versions proposed by refreshVersions?
If you find yourself deleting them every time you run the task, read on.
We are thinking about adding the ability to filter out dependencies.

One way would be with a **rejectVersionIf** predicate that would have access to the artifact coordinates (group and name), the version key (e.g. `version.moshi`), and the version and stability level of the current version and of the version candidate.

Here are a few examples

**Reject versions that are less stable than the current version**

```kotlin
refreshVersions {
    rejectVersionIf {
          candidate.stabilityLevel.isLessStableThan(current.stabilityLevel)
    }
}
```

**Use only stable version**

```kotlin
refreshVersions {
    rejectVersionIf {
          candidate.stabilityLevel != Stable
    }
}
```

**Use an allow list and a block list**

```kotlin
refreshVersions {
  rejectVersionIf {
        val allowList = listOf("version.kotlinx.serialization")
        val blockList = listOf("version.moshi")
        when {
            versionKey in blockList -> true
            candidate.stabilityLevel == Stable -> versionKey !in allowList
            else -> false
        }
    }
 }
```

Another way would be to define the default stability level, and add contextual metadata comments to override it:

```properties
#### Generated by `./gradlew refreshVersions` version x.x.x
#### ...
## defaultStabilityLevel = Alpha
```

```properties
## minimumum stability level: Beta
plugin.com.squareup.sqldelight=1.5.0
```

```properties
## minimumStabilityLevel = Stable
version.androidx.test=1.3.0

## minimumStabilityLevel = Stable
version.androidx.test.core=1.3.0

## minimumStabilityLevel = Stable
version.androidx.test.espresso=3.3.0

## minimumStabilityLevel = Stable
version.androidx.test.ext.junit=1.1.2
 ```

**Do you have use cases for filtering versions? Any help would be appreciated**

**Comment by colmoflynn** [comment]
Hi @jmfayard 

I have a question relating to the above. 

I have a beginner to intermediate level of knowledge of Gradle. I have only recently started to use the `refreshVersions` plugin. 
So, if my question has been answered elsewhere in the plugin documentation, or is based on a misunderstanding of how Gradle works or how this plugin works, my apologies in advance.

I have found that I can only declare the `refreshVersions` plugin dependency in `settings.gradle` and also to declare the `rejectVersionIf` predicate in the `settings.gradle`. 

Is there a way in the `rejectVersionIf` predicate to call logic that is project specific?.

For example: In the example cited here  https://docs.gradle.org/current/userguide/multi_project_builds.html , there is a single `settings.gradle` that can be shared between multiple projects. 

I was wondering if there was a means of calling a function that could pull back different filtering criteria for a different projects?.

e.g.  Could `rejectVersionIf` look for a function called `getBlockList()` that would return a list of module group IDs , that return one list for _projectA_ and a different list for _projectB_?. 

Thanks in advance.

**Comment by jmfayard** [comment]
No there isn't afaik, the same filter is applied to all projects

Le mar. 23 nov. 2021 à 19:25, colmoflynn ***@***.***> a
écrit :
